core/src/main/java/cucumber/runtime/Runtime.java

```diff
@@ -40,37 +38,29 @@ public Runtime(Plugins plugins,
         this.runtimeOptions = runtimeOptions;
         this.filters = filters;
         this.bus = bus;
-        this.runner = runnerSupplier.get();
+        this.runnerSupplier = runnerSupplier;
         this.featureSupplier = featureSupplier;
         exitStatus.setEventPublisher(bus);
     }
 
     public void run() {
         List<CucumberFeature> features = featureSupplier.get();
         bus.send(new TestRunStarted(bus.getTime()));
-        for (CucumberFeature feature : features) {
-            feature.sendTestSourceRead(bus);
-        }
-
+        
         StepDefinitionReporter stepDefinitionReporter = plugins.stepDefinitionReporter();
-
-        runner.reportStepDefinitions(stepDefinitionReporter);
-
-        for (CucumberFeature cucumberFeature : features) {
-            runFeature(cucumberFeature);
+        runnerSupplier.get().reportStepDefinitions(stepDefinitionReporter);
+        
+        final int requestedThreads = runtimeOptions.getThreads();
+        if (!features.isEmpty()) {
+            for (CucumberFeature feature : features) {
```

Comment:
Why was this moved?

Reply:
well I deliberately check to see if there are any features to be run, before running them.

it seemed strange to try and loop over something that could be empty, only a few lines above my check

Reply:
It changes the order of the events. Now the test source read events may be interlaced with other events.

Reply:
well all the existing tests passed, and that event won't be queued up,  so it should be a problem.

if you have concerns though I'll move it back

Reply:
Yes please. It is one less change in an implicit protocol.

Reply:
OK, again will wait for all feedback and then do all the changes in one final commit